Code/src/megadesk.cpp

```diff
@@ -912,53 +895,88 @@ uint8_t linBurst()
       break;
     }
     break;
+
   case State::UP:
-    if (user_cmd != Command::UP || currentHeight >= maxHeight)
+    enc_target = enc_min;
+    lin_cmd = LIN_CMD_RAISE;
+    lastState = State::UP;
+    if (user_cmd != Command::UP || enc_max >= maxHeight)
     {
       state = State::STOPPING1;
     }
     break;
+
   case State::DOWN:
-    if (user_cmd != Command::DOWN || currentHeight <= minHeight)
+    enc_target = enc_max;
+    lin_cmd = LIN_CMD_LOWER;
+    lastState = State::DOWN;
+    if (user_cmd != Command::DOWN || enc_min <= minHeight)
     {
       state = State::STOPPING1;
     }
     break;
+
   case State::STOPPING1:
-    state = State::STOPPING2;
-    break;
   case State::STOPPING2:
-    state = State::STOPPING3;
-    break;
   case State::STOPPING3:
-    state = State::STOPPING4;
+    enc_target = targetHeight; // more accurate stops
+    lin_cmd = LIN_CMD_FINE;
+    if (state == State::STOPPING1)
+      state = State::STOPPING2;
+    else if (state == State::STOPPING2)
+      state = State::STOPPING3;
+    else // State::STOPPING3
+      state = State::STOPPING4;
     break;
+
   case State::STOPPING4:
-    if (node_a[2] == node_b[2])
+    if (lastState == State::UP)
+      enc_target = enc_min;
+    else
+      enc_target = enc_max;
+    lin_cmd = LIN_CMD_FINISH;
+    // only check one, as that's good enough...
+    if (isIdle(node_a[2]))
     {
-      if (node_a[2] == LIN_MOTOR_IDLE1 || node_a[2] == LIN_MOTOR_IDLE2 || node_a[2] == LIN_MOTOR_IDLE3)
-      {
-        state = State::OFF;
-      }
+      state = State::OFF;
     }
+#if (defined SERIALCOMMS && defined SERIAL_IDLE)
+    writeSerial(response_stop, node_a[2], node_b[2], idleMarker); // Indicate the idle variant type
+#endif
     break;
+
   // recal stuff here
   case State::STARTING_RECAL:
+    lin_cmd = LIN_CMD_PREMOVE;
     state = State::RECAL;
     break;
+
   case State::RECAL:
-    if (node_a[0] <= 99 && node_a[1] == 0 && node_a[2] == 1 &&
-        node_b[0] <= 99 && node_b[1] == 0 && node_b[2] == 1) // Are both motors reporting 99/0/1? We should be bottomed out at this point.
+    // We want to send 0/0/189
+    enc_target = 0;
+    lin_cmd = LIN_CMD_RECALIBRATE;
+    if (enc_max <= 99 && node_a[2] == 1 && node_b[2] == 1) // Are both motors reporting 99/0/1? We should be bottomed out at this point.
         state = State::END_RECAL;
     break;
+
   case State::END_RECAL:
+    // We want to send 99/0/188
+    enc_target = 99;
+    lin_cmd = LIN_CMD_RECALIBRATE_END;
     state = State::OFF;
+    targetHeight = enc_max; // prevents immediately resuming previous height
```

Comment:
Leaving this line out might be considered a feature - desk would then return to the previous height after a recalibration.
But I figure it's more useful to be able to leave it at the absolute lowest height like this.